### PR TITLE
Check for dead/broken URLs (disabled by default)

### LINKED ISSUE #2128: Test for broken links
From the discussion on #1990, it'd be nice to have automatic checks on URLs that appear in the BCD data, both in `mdn_urls` and in notes.

I have a little bit of a proof of concept in the form of [a script I use](https://gist.github.com/ddbeck/290eadfa4cf40a0c34fc34b08d5fa28d) as a local pre-commit hook. Unfortunately, it's not at all clever, uses some non-standard tools, and it's a fish shell script, so it's not exactly portable. There's probably a right way to do this but I'm not sure what that would look like.

Additionally, previously related issues #1998 and #1703 imply that it may be desirable to have some broken links (e.g., MDN pages that _should_ exist, but don't).

### PR DESCRIPTION
**Pull request by queengooborg**
This PR introduces a new link test that checks for dead and/or broken URLs, thus this PR fixes #2128.

*WARNING:* This new test is SUPER slow due to how many links there are.  (As in, it took roughly 30 minutes on Gigabit internet just to get through the API folder, and an hour to complete the whole process.)  I've added a variable so we can toggle it.  This variable will be off by default.

### CODE REVIEW

**Comment by vinyldarkscratch** [comment]
Wow, good catch @sideshowbarker!  I will see what I can do to get that fixed.  :wink:  (Currently, I will be directing my primary focus towards Q4 Sprint 4, AKA getting all the real versions for JavaScript data for Safari, so it'll be a little while; not like this PR will be merged anytime soon since it builds off of another.  😛)

File: test/linter/test-links.js [hunk 1]
```diff
@@ -120,77 +118,87 @@ function processData(filename, logger) {
...
+      return {'issue': 'Use non-localized Microsoft Developer URL', 'expected': `https://developer.microsoft.com/${match[2]}`};
     }
   );
 
-  return hasErrors;
+  if (testDeadLinks) {
+    processLink(
+      errors,
+      actual,
+      String.raw`(https?):\/\/((www\.)?[-a-zA-Z0-9@:%._\+~#=]{1,256}\.[a-zA-Z0-9()]{1,6})\b([-a-zA-Z0-9()@:%_\+.~#?&//=]*)`,
```
Comment (sideshowbarker) [inline test/linter/test-links.js]:
```suggestion
      String.raw`(https?):\/\/((www\.)?[-a-zA-Z0-9@:%._\+~#=]{1,256}\.[a-zA-Z0-9()]{1,6})\b([-a-zA-Z0-9()@:%_\+.~#?&//=*]*)`,
```

**Review by alshamiri5 — APPROVED** [review]

**Comment by vinyldarkscratch** [comment]
Hi @alshamiri5, I see you've been adding approval reviews to lots of PRs in this repository.  If you'd like to find out how you can contribute to this project, please check out https://github.com/mdn/browser-compat-data/blob/master/docs/contributing.md, as well as issues labeled as "help wanted"!

**Comment by Elchi3** [comment]
Thanks for your PR and sorry there was no conclusion here sooner.

I'm going to close this now because the 404 mdn_url discussion is happening in https://github.com/mdn/browser-compat-data/issues/11291. Also, https://github.com/mdn/yari/issues/5015 has been filed to produce a content inventory that should make querying MDN for existing content much faster. That would be the preferred approach over this PR. Thanks for doing this nevertheless. It inspired a more robust solution and a useful discussion.

**Comment by queengooborg** [comment]
Oh sweet, glad to hear this is being worked on in Yari now!  It'll definitely be a much less ugly-performing solution than this!  😝

**Comment by queengooborg** [comment]
I saw the related issue to this once again, and I realized that this problem spans to more than just MDN URLs.  While MDN URLs are a big part of this, we also have other links in BCD that are probably more prone to 404-ing, such as spec URLs, and more importantly, URLs inside of notes.  I think that we should revisit this PR and research a way to validate URLs are still alive in BCD.

**Pull request closed without merging**